### ISSUE TITLE
onready var node = requires(nodepath)

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#2618]
### Describe the project you are working on
Spaceship game

### Describe the problem or limitation you are having in your project
Scripts can be attached to basically anything even if it couldn't run due to missing nodes in the tree

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Allow a requires statement where we can load nodes that are required in the scene tree

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
```
extends KinematicBody2d

onready var Collider = requires("Collider")
#or
requires("Collider") 
```

This would disallow the script to be attached to the node and throw a error if these conditions are not met.

You could even go as far as not implementing requires and just requiring all get_node(...) and $... statements to succeed before the script can be attached, but that's if people support it I guess.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
It would be used all the time to ensure that the script can run successfully node dependency errors
also it can be worked around but not everyone is going to ensure their script meets these standards

### Is there a reason why this should be core and not an add-on in the asset library?
See above.

**Comment by aaronfranke** [comment]
The ability to export nodes is already tracked [here](https://github.com/godotengine/godot-proposals/issues/1048). Currently, the best option is to export NodePaths.

> This would disallow the script to be attached to the node and throw a error if these conditions are not met.

This will not be done. If you don't want a script attached in a place you don't want it, just don't add the script there. You can also add checks to the script to ensure that it has the correct children at runtime.

**Comment by ka23ppa3** [comment]
Thats reminds me about Odin Inspector plugin for Unity, which help to determine and solve problems **before the project run**. In C#/Unity with Odin it looks like:

```
[Required]
public Transform transformNode;
```
So attribute ```[Required]``` tracks if ```transformNode``` is assigned from the editor or not and if not it will throw the error message. If you use Odin Inspector + OdinValidator the last one will open separated window with all mistakes, missing require fields etc.

https://odininspector.com/attributes/required-attribute
https://odininspector.com/files/pages/odin-project-validator/odin-validator-example-usecase.png

**PLEASE** keep in mind that when we working with editor we need more of such "control-attributes" on one field and they need to stack with each others.

**Comment by ka23ppa3** [comment]
> > This would disallow the script to be attached to the node and throw a error if these conditions are not met.

I disagree to block it to attach, but instead just show error that field are required to be initialized from editor before launch.

> This will not be done. If you don't want a script attached in a place you don't want it, just don't add the script there. You can also add checks to the script to ensure that it has the correct children at runtime.

Of course we can to track those errors in runtime but I already can say that we will lost a lot of time because of this, before the error popup. Just because we cannot test everytime whole game, spawn all nodes at the scene etc to make sure all checks was passed.

**Comment by Jummit** [comment]
A "required" parameter to exported node paths does indeed sound useful:
`@export_node_path(Sprite, true)` would show a warning if no sprite is assigned.
